lib/transfer.c

```diff
@@ -1286,6 +1287,75 @@ void Curl_init_CONNECT(struct Curl_easy *data)
   data->state.in = data->set.in_set;
 }
 
+#if defined(WIN32)
+static CURLcode check_for_option_conflicts(struct Curl_easy *data)
+{
+  if(Curl_GetModuleHandleExA) {
+    HMODULE module;
+
+/* disable warning for function pointer to data pointer */
+#if defined(__GNUC__)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wpedantic"
+#elif defined(_MSC_VER)
+#pragma warning(push)
+#pragma warning(disable:4054)
+#endif
+    bool dll = /* true if our code is running from a DLL */
+      Curl_GetModuleHandleExA(
+        (GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS |
+         GET_MODULE_HANDLE_EX_FLAG_UNCHANGED_REFCOUNT),
+        ((LPCSTR)(void *)check_for_option_conflicts), &module) &&
+      module != GetModuleHandleA(NULL);
+#if defined(__GNUC__)
+#pragma GCC diagnostic pop
+#elif defined(_MSC_VER)
+#pragma warning(pop)
+#endif
+
+    if(dll) {
+      if(data->set.is_ioctl_func_from_user &&
+         (!data->set.is_in_set_from_user ||
+          !data->set.is_fread_set_from_user)) {
+        failf(data, "Windows DLL users must set CURLOPT_READDATA and "
+                    "CURLOPT_READFUNCTION if CURLOPT_IOCTLFUNCTION is set.");
+        return CURLE_BAD_FUNCTION_ARGUMENT;
+      }
+      if(data->set.is_seek_func_from_user &&
+         (!data->set.is_in_set_from_user ||
+          !data->set.is_fread_set_from_user)) {
+        failf(data, "Windows DLL users must set CURLOPT_READDATA and "
+                    "CURLOPT_READFUNCTION if CURLOPT_SEEKFUNCTION is set.");
+        return CURLE_BAD_FUNCTION_ARGUMENT;
+      }
+      if(data->set.is_in_set_from_user &&
+         !data->set.is_fread_set_from_user) {
+        failf(data, "Windows DLL users must set CURLOPT_READFUNCTION if "
+                    "CURLOPT_READDATA is set.");
+        return CURLE_BAD_FUNCTION_ARGUMENT;
+      }
+      if(data->set.is_out_from_user &&
+         !data->set.is_fwrite_func_from_user) {
+        failf(data, "Windows DLL users must set CURLOPT_WRITEFUNCTION if "
+                    "CURLOPT_WRITEDATA is set.");
+        return CURLE_BAD_FUNCTION_ARGUMENT;
+      }
+      if(data->set.is_writeheader_from_user &&
+         !data->set.is_fwrite_func_from_user &&
+         !data->set.is_fwrite_header_from_user) {
+        failf(data, "Windows DLL users must set CURLOPT_HEADERFUNCTION or "
+                    "CURLOPT_WRITEFUNCTION if CURLOPT_HEADERDATA is set.");
+        return CURLE_BAD_FUNCTION_ARGUMENT;
+      }
+    }
+  }
+
+  return CURLE_OK;
```

Comment:
I think you could simplify this logic and still cover 99% of the mistakes: If the `*DATA*` argument is not set, but the `*FUNCTION*` is, that's the main problematic case. You can do that without introducing any `is_*_from_user` bits.